server/src/main/java/org/elasticsearch/cluster/metadata/TemplateUpgradeService.java

```diff
@@ -124,7 +122,8 @@ public void clusterChanged(ClusterChangedEvent event) {
         lastTemplateMetaData = templates;
         Optional<Tuple<Map<String, BytesReference>, Set<String>>> changes = calculateTemplateChanges(templates);
         if (changes.isPresent()) {
-            if (updatesInProgress.compareAndSet(0, changes.get().v1().size() + changes.get().v2().size())) {
+            if (upgradesInProgress.compareAndSet(0, changes.get().v1().size() + changes.get().v2().size() + 1)) {
```

Comment:
`+ 1` so that the `calculateTemplateChanges` check after the upgrade, is also guarded by this `compareAndSet` condition.